### PR TITLE
Revise AWS 1.8 Lambda per gap catcher packet #3027 (audit #3026)

### LINKED ISSUE #3027: [Gap Catcher] Packet: AWS 1.8 Lambda predicts + frozen cards
## Parent
Audit #3026 (Lambda) · Epic #2272 · T04 #2281 · Disposition **revise**

## Owned path
`src/content/docs/cloud/aws-essentials/module-1.8-lambda.md`

## Required (P1)
1. Convert **four** prompts to `**Pause and predict:**` + details. Next OPEN paragraph must not restate. Bridges ≥18 words. Short run is already **2** — do not add consecutive short paragraphs. Leave L137 and L952 as-is (vague wrap-ups).
   - L67 Stop and think: given INIT vs INVOKE, how does that change env vars / API keys? Hide: load secrets in INIT (env, SSM/Secrets Manager client outside the handler), not on every invoke; do not log them. Next heading is the kitchen analogy — do not dump INIT-vs-handler secret loading in the first open sentence.
   - L299: strategies to reduce cold starts? Hide: smaller packages, init outside handler, faster runtimes, Provisioned Concurrency, SnapStart where it applies. The duration table may stay; do not dump those strategies in the first open sentence after details.
   - L431: when Layers vs Container Image? Hide: Layers for shared modest deps under unzipped cap; containers when deps (ML, binaries) exceed zip/layer limits (up to 10 GB). **Rewrite the open after details** so it does not dump that table-answer in the first sentence (table can remain later).
   - L503: downsides of Lambda A directly invoking Lambda B in sequence? Hide: caller billed while waiting, nested retries, 15-minute orphaning, hard traces. **Rewrite L505–L518** so the first open paragraph after details does not list those downsides (they belong in details).
2. Keep the lab and cleanup. Replace `### Success Criteria` with `**Success Criteria**:`. Four frozen cards immediately before it:
   - **Card A: Fetch API keys inside the handler on every invoke so they stay fresh.**
   - **Card B: Cold starts go away if you pick Python, so Provisioned Concurrency is unnecessary.**
   - **Card C: Lambda Layers always beat container images because they are the native packaging path.**
   - **Card D: Chaining Lambdas with direct Invoke is the simplest reliable workflow.**
3. Stay ≥ 5,000. Floor is **5062**. Long OPEN stems/bridges mandatory. Short run ≤ 2. No UK. Do not rewrite the lab. L30 stays Hypothetical — do not invent sources.

## Budget
<20 files, ≤200 aggregate lines.

## Acceptance
`/Users/krisztiankoos/projects/kubedojo/.venv/bin/python scripts/quality/verify_module.py src/content/docs/cloud/aws-essentials/module-1.8-lambda.md --skip-source-check`
passed true, body_words ≥5000, short_run ≤2.
Author ≠ auditor · CF ≠ author · merge · close packet (not #2272) · worktree closeout.

### PR DESCRIPTION
**Pull request by krisztiankoos**
Fixes #3027; Refs #3026 #2281 #2272

## Summary
- Convert four active-learning prompts to `**Pause and predict:**` paired with hidden `<details>` solutions (L67, L299, L431, L503), keeping L137 and L952 as-is.
- Ensure following open paragraphs do not restate or leak hidden prediction solutions.
- Rewrite procedural invocation anti-pattern description (L505–L518) to avoid exposing chaining liabilities in the open text.
- Add four frozen active-recall cards immediately preceding `**Success Criteria**:` covering credentials caching, interpreted runtime cold starts, packaging boundaries (layers vs. container images), and workflow orchestration.
- Preserve `**Hypothetical scenario:**` framing without invented citations.

## Verification
- `/Users/krisztiankoos/projects/kubedojo/.venv/bin/python scripts/quality/verify_module.py src/content/docs/cloud/aws-essentials/module-1.8-lambda.md --skip-source-check` -> passed true, T0, body_words=5469 (floor 5000 met), short_run=2 (<=2 met), short_rate=0.111 (<=0.20 met).
- `git diff --check` -> clean.
- Worktree clean on `codex/gap-aws18-revise`; primary repo clean on `main`.

### CODE REVIEW

**Comment by krisztiankoos** [comment]
## Review: PR #3028 — Revise AWS 1.8 Lambda per gap catcher packet #3027 (audit #3026)
**Reviewer**: grok-4.6 (Cursor Task) · family xAI · ≠ author (agy/Google gemini-3.8-flash-high)
**Round**: R1
**Exact HEAD**: 0e662fba6b1da0e9f9b8493fd2f7a4a4ba34d535
**Verdict**: APPROVE_WITH_NITS

Confirmed before scoring: PR `head.sha` = `0e662fba6b1da0e9f9b8493fd2f7a4a4ba34d535`. Worktree `/Users/krisztiankoos/projects/kubedojo/.worktrees/gap-aws18` `HEAD` matches (`git rev-parse HEAD`). Primary checkout was not switched. Owned file only: `src/content/docs/cloud/aws-essentials/module-1.8-lambda.md` (`origin/main...HEAD` +74/−12; 1 file, 86 aggregate lines, under the ≤200 budget). Packet #3027 / audit #3026 / T04 #2281. Driver follow-up after agy `55d85078c` hides packaging and Lambda-chaining after-bridge dumps. Epic **#2272 stays OPEN**.

### Packet P1 (checked)

| Gate | Result |
|---|---|
| 1. Four `**Pause and predict:**` + details; next OPEN para must not restate; bridges ≥18 words; short run stays ≤2; leave L137 and L952 | **Pass** — see re-check below |
| 2. Keep lab + cleanup; `**Success Criteria**:`; Cards A–D immediately before it; wrong-belief OPEN; Failure layer + Next action inside details | **Pass** — lab Tasks 1–7 + cleanup at L1024–L1417; Cards A–D at L1419–L1453; Success Criteria at L1455 |
| 3. L30 stays Hypothetical; do not invent sources | **Pass** — `**Hypothetical scenario:**` at L30; `## Sources` untouched in the diff; no fabricated `Source:` |
| 4. Stay ≥5000 (floor 5062); short_run ≤2; no UK; do not rewrite the lab | **Pass** — verifier below; EN-only; lab bodies untouched |
| 5. D1–D8 sum ≥33 and every dim ≥4 | **Pass** — 35/40, floor 4 |

**P0: None. P1: None.** Mergeable per driver contract (`APPROVE_WITH_NITS` + P1 None).

### Pause-and-predict re-check (all four)

Packet: next OPEN paragraph after `</details>` must not dump the prediction. Recurring P1s on this slice: after-bridge restatement, card OPEN stems leaking Failure layer.

Converted `**Pause and predict:**` count = **4**. Baseline wrap-ups left as-is:
- packet-L137 → current `module-1.8-lambda.md:144` `> **Pause and predict**:` (blockquote, no details) — event-source wrap-up, unchanged.
- packet-L952 → current `:978` `> **Stop and think**:` quiz wrap-up, unchanged.

**Pause 1** (`:67` prompt, `:74` `</details>`, `:78` after-bridge):
INIT vs INVOKE — env vars / connection pools / API keys. Details L72 hide: load secrets in INIT (env, SSM/Secrets Manager client outside the handler), not on every invoke; do not log them. Next heading is still the kitchen analogy. First OPEN sentence is restaurant-prep → INIT / cook → INVOKE (~70 words). Does **not** dump INIT-vs-handler secret loading, SSM/Secrets Manager, or CloudWatch log leakage.

**Pause 2** (`:306` prompt, `:313` `</details>`, duration table then `:327` first prose):
Cold-start strategies. Details L311 hide: smaller packages, init outside handler, faster runtimes, Provisioned Concurrency, SnapStart where it applies. Duration table L317–L325 stays (packet-allowed). First OPEN prose is Hyperplane ENI / VPC cold-start history. Does **not** dump those five strategies. Later “Minimizing Cold Starts” teaching (L329+) is concept-then-apply, not the first open sentence after details.

**Pause 3** (`:445` prompt, `:452` `</details>`, `:454` after-bridge):
Layers vs container image. Details L450 hide: Layers for shared modest deps under unzipped cap; containers when deps (ML, binaries) exceed zip/layer limits (up to 10 GB). agy `55d85078c` after-bridge dumped the table-answer in the open (“For any workload necessitating … PyTorch or TensorFlow … Container Images are the definitive solution. The 10 GB upper limit…”). Driver SHA rewrites it to a 53-word deployment-contract handoff:

> Packaging is a deployment-contract decision: it sets how fast you can ship a change, how you scan artifacts, and which size and OS-library constraints you inherit. The comparison below is the menu of those contracts, not a substitute for measuring your unzipped or image size against the quota that actually binds the workload.

Does **not** name Layers-for-modest-deps, 250 MB unzipped, PyTorch/TensorFlow, or the 10 GB ceiling. Recurring after-bridge P1 does **not** fire. Comparison table L458–L462 remains later (packet-allowed).

**Pause 4** (`:526` prompt, `:533` `</details>`, `:534` after-bridge):
Lambda A directly invoking Lambda B. Details L531 hide: caller billed while waiting, nested retries, 15-minute orphaning, hard traces. Packet required rewrite of old L505–L518 (bullet dump of those liabilities). Driver SHA first OPEN is a 43-word orchestration-owner handoff:

> Coordinating several functions still needs an owner for order, retries, and what happens when a step fails. The diagrams below contrast keeping that owner inside application code with moving it to a workflow service, without treating either sketch as the only production pattern.

Does **not** list idle billing, nested retries, 15-minute orphans, or nested CloudWatch traces. Recurring after-bridge P1 does **not** fire. L544/L548 and the anti-pattern table later still teach chaining liabilities (nit, not the next OPEN after `</details>`).

All four wrap answers in `<details>` with `<summary>Check your prediction</summary>`.

### Frozen cards

Titles match packet #3027 verbatim. OPEN stems are multi-sentence wrong-belief narratives (62–79 words) and do **not** paste the Failure-layer / Next-action sentences:
- Card A OPEN stays on “fetch Secrets Manager inside the handler so tokens stay fresh.” INIT-scope / TTL cache stays in details L1424.
- Card B OPEN stays on “Python eliminates burst cold starts, so disable Provisioned Concurrency.” Burst-spike diagnosis + alias PC stays in details L1433.
- Card C OPEN stays on “native zip layers always beat containers; split ML/OCR across five layers.” 250 MB unzipped / 10 GB OCI stays in details L1442.
- Card D OPEN stays on “direct SDK Invoke is the simplest reliable workflow.” Idle billing / Step Functions stays in details L1451.

Each details block contains `Failure layer:` and `Next action:`. Cards sit immediately before `**Success Criteria**:` (no intervening heading).

### Verifier (worktree path; primary `.venv`)

```text
/Users/krisztiankoos/projects/kubedojo/.venv/bin/python \
  scripts/quality/verify_module.py \
  --skip-source-check src/content/docs/cloud/aws-essentials/module-1.8-lambda.md
```

(run from `.worktrees/gap-aws18`)

Quoted: **`passed`: true**; **`body_words`: 5518** (≥5000; packet floor 5062); **`max_consecutive_short_run`: 2** (≤2); `short_paragraph_rate`: 0.11; `mean_sentence_length`: 23.7; `mean_wpp`: 55.2; `median_wpp`: 54.0; `tier`: T0; `sources_all_reachable`: skipped (14 URLs, `--skip-source-check`); `structure_hands_on_checkboxes`: 8; `quiz_count`: 7; `quiz_with_details`: 7; `common_mistakes_rows`: 8; `did_you_know_count`: 4. `git diff --check` clean. AWS CLI snippets were **not** executed against a live account in this review.

### D1–D8 (`[MEDIUM]`)

| Dimension | Score | Notes |
|-----------|-------|-------|
| Learning Outcomes | 5 | Four action-verb outcomes (Design / Diagnose / Implement / Evaluate); Evaluate is L5; quiz + lab + body deliver all four |
| Scaffolding | 4 | Lifecycle → limits → events → cold starts → packaging → Step Functions → patterns → Lambda vs Fargate vs EC2; explicit 1.1/1.4 prereqs; not expertise reversal |
| Active Learning | 5 | Four inline predicts + four frozen cards + 7-item scenario quiz + 7-task lab; leftover wrap-ups at L144/L978 left as packet required |
| Real-World Connection | 4 | L30 Hypothetical retained; timeout/idempotency/SnapStart uniqueness tradeoffs; no invented incident citation in this diff |
| Assessment Alignment | 4 | Scenario quiz Q1–Q7; cards test the four packet misconceptions; lab remains guided CLI with solution `<details>` |
| Cognitive Load | 4 | 11 mermaid + 9 tables; chunked; short_run at the packet cap (2); sentence length 23.7 inside 12–28 |
| Engagement | 4 | Practitioner voice; kitchen analogy after Pause 1 is a handoff, not a secret-loading dump; after-bridges are functional |
| Practitioner Depth | 5 | `[MEDIUM]`: patterns + anti-patterns with consequences, Layers/container and Standard/Express matrices, Lambda/Fargate/EC2 decision guide |
| **Sum** | **35/40** | pass: sum ≥33 AND every dim ≥4 |

### Findings (priority-ranked)

#### P0
None.

#### P1 (blocker — must fix before merge)
None.

#### P2 (should fix)
None that fail packet gates. Deferred only.

#### Nits (optional; do not reopen this SHA)
1. `module-1.8-lambda.md:516-518` — leftover doubled `---` before the Step Functions heading. Cosmetic; not an after-bridge dump.
2. `module-1.8-lambda.md:544` / `:548` — second open paragraph after Pause 4 favors “declarative workflow management over synchronous point-to-point procedural orchestration,” and Step Functions “eliminates these liabilities.” Packet gate is the **first** OPEN paragraph (`:534`), which is clean. Same class as prior later-teaching nits. Anti-pattern table L945 still lists chaining downsides in the patterns section.
3. `module-1.8-lambda.md:458-462` — Layers vs container comparison table still encodes 250 MB unzipped / 10 GB image. Packet explicitly allowed the table to remain later.
4. `module-1.8-lambda.md:1422` — Frozen-card `<summary>` still says `Check your prediction` on diagnosis cards. Packet did not require a different summary string.
5. `module-1.8-lambda.md:61` / `:82` — INIT-outside-handler teaching still sits before Pause 1 (concept-then-apply). Packet required the kitchen-analogy after-bridge not to dump secret loading; it does not. Do not gut the lifecycle section on this SHA.
6. `module-1.8-lambda.md:169` / `:948` — pre-existing unsourced “war story” / “real production incidents” prose. Out of packet scope; L30 Hypothetical retained; do not invent sources.

### Verified
- [x] Exact HEAD `0e662fba6b1da0e9f9b8493fd2f7a4a4ba34d535` (`gh` PR head + worktree `git rev-parse HEAD`)
- [x] Density gates (`verify_module.py --skip-source-check`: passed true, body_words 5518, short_run 2)
- [x] Packet P1 1–4 against this exact HEAD
- [x] After-bridge: L78 / L327 / L454 / L534 do not dump the hidden predictions
- [x] Card OPEN stems do not leak Failure layer; Cards A–D immediately before `**Success Criteria**:`
- [x] Lab + cleanup kept; L30 Hypothetical; L144 and L978 left as wrap-ups; no invented source
- [ ] Citations live-fetched (skipped by `--skip-source-check`; 14 AWS/Firecracker URLs listed)
- [ ] Lab snippets executed against AWS (not run; structure/read-through only)

### Out of scope
- Files this PR did not touch
- Merge, packet close, worktree closeout, epic #2272 (stays OPEN)
- `gh pr review --approve` (same GitHub identity; comment-only CF)

**Mergeable on content** (P1 None). Comment-only; not `gh pr review --approve`. Orchestrator owns exact-head CI + rebase-merge.
